storm-core/src/native/worker-launcher/impl/utils/file-utils.c

```diff
@@ -0,0 +1,182 @@
+  int bytes_read;
+  while ((bytes_read = read(fd, cp, bytes_left)) > 0) {
+    cp += bytes_read;
+    bytes_left -= bytes_read;
+    if (bytes_left == 0) {
+      buff_size += FILE_BUFFER_INCREMENT;
+      bytes_left += FILE_BUFFER_INCREMENT;
+      buff = realloc(buff, buff_size);
+      if (buff == NULL) {
+        fprintf(ERRORFILE, "Unable to allocate %ld bytes\n", buff_size);
+        goto cleanup;
+      }
+    }
+  }
+  if (bytes_left < 0) {
+    fprintf(ERRORFILE, "Error reading %s : %s\n", filename, strerror(errno));
+    goto cleanup;
+  }
+
+  *cp = '\0';
+  rc = 0;
+
+cleanup:
+  if (fd != -1) {
+    close(fd);
+  }
+  if (rc != 0) {
+    free(buff);
+    buff = NULL;
+  }
+  return buff;
+}
+
+/**
+ * Read a file to a string as the worker-launcher user and returns the
+ * result as a string. See read_file_to_string for more details.
+ *
+ * Returns a pointer to the allocated, NUL-terminated string or NULL on error.
+ */
+char* read_file_to_string_as_wl_user(const char* filename) {
+  uid_t user = geteuid();
+  gid_t group = getegid();
+  if (change_effective_user_to_wl() != 0) {
+    fputs("Cannot change to worker-launcher user\n", ERRORFILE);
+    return NULL;
+  }
+
+  char* buff = read_file_to_string(filename);
+  if (change_effective_user(user, group) != 0) {
+    fputs("Cannot revert to previous user\n", ERRORFILE);
+    free(buff);
+    return NULL;
+  }
+  return buff;
+}
+
+/**
+ * Write a sequence of bytes to a new file as the worker-launcher user.
+ *
+ * Returns true on success or false on error.
+ */
+bool write_file_as_wl(const char* path, const void* data, size_t count) {
+   bool result = false;
+  int fd = -1;
+  uid_t orig_user = geteuid();
+  gid_t orig_group = getegid();
+  if (change_effective_user_to_wl() != 0) {
+    fputs("Error changing to worker-launcher user and group\n", ERRORFILE);
+    return false;
+  }
+
+  fd = open(path, O_CREAT | O_EXCL | O_WRONLY, S_IRUSR | S_IWUSR);
+  if (fd == -1) {
+    fprintf(ERRORFILE, "Error creating %s : %s\n", path, strerror(errno));
+    goto cleanup;
+  }
+
+  const uint8_t* bp = (const uint8_t*)data;
+  while (count > 0) {
+    ssize_t bytes_written = write(fd, bp, count);
+    if (bytes_written == -1) {
+      fprintf(ERRORFILE, "Error writing to %s : %s\n", path, strerror(errno));
+      goto cleanup;
+    }
+    bp += bytes_written;
+    count -= bytes_written;
+  }
+
+  result = true;
+
+cleanup:
+  if (fd != -1) {
+    if (close(fd) == -1) {
+      fprintf(ERRORFILE, "Error writing to %s : %s\n", path, strerror(errno));
+      result = false;
+    }
+  }
+
+  if (change_effective_user(orig_user, orig_group) != 0) {
+    fputs("Cannot restore original user/group\n", ERRORFILE);
+    result = false;
+  }
+
+  return result;
+}
+
+char *get_full_path(const char *dir, const char* file) {
+  char *ret;
+  int bytesPrinted = asprintf(&ret, "%s/%s", dir, file);
+  if (bytesPrinted == -1) {
+    fprintf(ERRORFILE, "ERROR: failed to allocate file path\n");
```

Comment:
Add info about exiting, like: "ERROR:EXITING process: failed to allocate....." because of the next exit() call.